Table view for search results

fiskus opened the pull request:
# Description

![Screenshot from 2025-06-09 16-53-41](https://github.com/user-attachments/assets/0591793c-74eb-47b0-8ebf-bcf1845becac)

# TODO

<!-- Remove items that are irrelevant to this PR -->

- [ ] Unit tests
- [ ] Automated tests (e.g. Preflight)
- [ ] Confirm that this change meets security best practices and does not violate the security model
- [ ] Documentation
    - [ ] run `optipng` on any new PNGs
    - [ ] [Python: Run `build.py`](../tree/master/gendocs/build.py) for new docstrings
    - [ ] JavaScript: basic explanation and screenshot of new features
    - [ ] Markdown somewhere in docs/**/*.md that explains the feature to end users (said .md files should be linked from SUMMARY.md so they appear on https://docs.quilt.bio)
    - [ ] Markdown docs for developers
- [ ] [Changelog](../tree/master/docs/CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

fiskus commented on shared/graphql/schema.graphql:
```diff
@@ -311,6 +311,8 @@ enum SearchResultOrder {
   NEWEST
   OLDEST
+  LEX_ASC
+  LEX_DESC
```
Thanks! I reverted the schema

drernie commented:
@greptile-app

fiskus commented on catalog/app/containers/Search/Layout/ScrollToTop.tsx:
```diff
@@ -0,0 +1,59 @@
+    position: 'fixed',
+    left: '50%',
+    bottom: t.spacing(3),
+    transform: `translateX(-50%)`,
+    animation: t.transitions.create('$slide'),
```
This is how MUI works, animation name should be prefixed with `$`

fiskus commented on catalog/app/containers/Bucket/Search.tsx:
```diff
@@ -0,0 +1,43 @@
+    model.state.resultType === SearchUIModel.ResultType.QuiltPackage
+  return (
+    <>
+      <MetaTitle>{['Packages', bucket, model.state.searchString || '']}</MetaTitle>
```
They can't be `undefined`. And we already check `searchString` for `null`, however, we shouldn't pass empty string as a list item to `MetaTitle,` because it renders it as ` •  • `

fiskus replied:
Fixed

fiskus commented on catalog/app/containers/Search/Table/Entries.tsx:
```diff
@@ -0,0 +1,338 @@
+            {entries.map((entry) => (
+              <Entry
+                className={classes.row}
+                key={entry.logicalKey + entry.physicalKey}
```
Yes, I used `physicalKey` first, and thought that it is not enough, so added `logicalKey`. But `logicalKey` itself is enough, thanks

fiskus commented on catalog/app/containers/Search/Table/CellValue.tsx:
```diff
@@ -0,0 +1,165 @@
+function UserMetaValue({ hit, pointer }: UserMetaValueProps) {
+  const value = React.useMemo(() => {
+    try {
+      const meta = hit.meta ? JSON.parse(hit.meta) : {}
```
Moved to `useColumns`, so `Hit` for this table view has `JsonRecord` as meta.
Though, maybe, makes sense to send meta as JSON from GQL?

nl0 replied:
> Though, maybe, makes sense to send meta as JSON from GQL?

in principle, yes, that should be the case, but there is a challenge wit that:
meta is returned as string from elastic, so in order to make it a dict, we have to parse it on the backend,
but then it will serialize it back to json, so that's significant amount of unnecessary work.
i'll try to come up with some efficient solution for this when i have some spare capacity

fiskus commented on catalog/app/containers/Bucket/Bucket.tsx:
```diff
@@ -0,0 +1,158 @@
+          <M.AppBar position="static" className={classes.appBar}>
+            <BucketNav bucket={bucket} section={section} />
+          </M.AppBar>
+          <M.Container
```
Yep, intended to use Layout/Container here, thanks

fiskus commented on catalog/app/containers/NavBar/Provider.tsx:
```diff
@@ -23,21 +19,17 @@ function useUrlQuery() {
 }
 
-function useSearchUIModel() {
+export function useSearchUIModel() {
```
Moved as `useSearchUIModelContextUnsafe` to `Search/model`

fiskus commented on catalog/app/containers/NavBar/Controls.tsx:
```diff
@@ -0,0 +1,149 @@
+  }, [setState])
+  React.useEffect(cancel, [cancel])
+
+  const selectRef: React.Ref<HTMLInputElement | void> = React.useRef()
```
Yes

-  const selectRef: React.Ref<HTMLInputElement | void> = React.useRef()
+  const selectRef: React.Ref<HTMLInputElement> = React.useRef(null)

fiskus commented on catalog/app/components/Layout/Container.tsx:
```diff
@@ -0,0 +1,48 @@
+
+const useContainerStyles = M.makeStyles((t) => ({
+  fullWidth: {
+    animation: t.transitions.create('$expand'),
```
MUI expects `$name`

fiskus commented on catalog/app/utils/JSONPointer.spec.ts:
```diff
@@ -18,4 +19,36 @@ describe('utils/JSONPointer', () => {
+    expect(jsonpath.value(obj, JSONPointer.toJsonPath(pointer))).toBe('true!')
+  })
+
+  it.skip('FIXME: should handle $$', () => {
+    const pointer = '/$$'
+    const obj = { $$: 'true!' }
+    expect(jsonpath.value(obj, JSONPointer.toJsonPath(pointer))).toBe('true!')
+  })
```
> without ticket reference

There is the `FIXME` note instead of a ticket

fiskus commented on catalog/app/containers/Search/Table/index.tsx:
```diff
@@ -0,0 +1,74 @@
+
+interface ResultsInnerProps {
+  className?: string
+  results: Extract<Results, { _tag: 'ok' }>
```
It makes sense here. But it also makes API for `useResults` more verbose. I would keep `exports` scarce

fiskus commented on catalog/app/containers/Search/Table/CellValue.tsx:
```diff
@@ -0,0 +1,165 @@
+    )
+  }
+
+  switch (typeof value) {
+    case 'number':
+    case 'string':
+      return <>{value}</>
+    case 'object':
+      return <>{JSON.stringify(value)}</>
+    default:
```
Re-used `JsonEditor/PreviewValue`

fiskus commented on catalog/app/containers/Search/Layout/ObjectFilters.tsx:
```diff
@@ -0,0 +1,151 @@
+  }, [deactivateObjectsFilter, field])
+
+  const change = React.useCallback(
+    (state: $TSFixMe) => {
+      setObjectsFilter(field, state)
```
Lazy workaround:
```diff
-    (state: $TSFixMe) => {
+    (state: Parameters<typeof setObjectsFilter>[1]) => {
```

fiskus commented on catalog/app/containers/Search/model.ts:
```diff
@@ -1203,6 +1234,30 @@ export const PackageUserMetaFacetTypeInfo = {
+          return undefined
+        case 'InvalidInput':
+          return undefined
+        case 'PackagesSearchResultSet':
+          return oneOf(['workflow', 'modified', 'size', 'entries'], field)
+            ? r.stats[field]
+            : undefined
+        default:
```
I don't think it's possible, but why not

fiskus commented on catalog/app/containers/Bucket/Bucket.tsx:
```diff
@@ -0,0 +1,158 @@
+
+  const { paths, urls } = NamedRoutes.use<RouteMap>()
+
+  const urlState: SearchUIModel.SearchUrlState = React.useMemo(
```
Added `makePackagesSearchUrlState/useMakePackagesSearchUrlState` to "Search/model.ts"

fiskus commented on catalog/app/containers/Search/Table/Provider.tsx:
```diff
@@ -0,0 +1,96 @@
+
+import type { Column } from './useColumns'
+
+export type HiddenColumns = Map<Column['filter'], boolean>
```
```diff
diff --git a/catalog/app/containers/Search/Table/Provider.tsx b/catalog/app/containers/Search/Table/Provider.tsx
index 42f089bd..ceaa0636 100644
--- a/catalog/app/containers/Search/Table/Provider.tsx
+++ b/catalog/app/containers/Search/Table/Provider.tsx
@@ -1,9 +1,10 @@
 import invariant from 'invariant'
 import * as React from 'react'
+import * as Eff from 'effect'
 
 import type { Column } from './useColumns'
 
-export type HiddenColumns = Map<Column['filter'], boolean>
+export type HiddenColumns = Eff.HashSet.HashSet<Column['filter']>
 
 interface Context {
   filter: Column | null
@@ -21,13 +22,6 @@ interface Context {
 
 const Ctx = React.createContext<Context | null>(null)
 
-const removeBool = (f: Column['filter']) => (x: HiddenColumns) => {
-  const m = new Map(x)
-  return m.delete(f) ? m : x
-}
-
-const addBool = (f: Column['filter']) => (x: HiddenColumns) => new Map(x).set(f, true)
-
 export function Provider({ children }: { children: React.ReactNode }) {
   const [filter, setFilter] = React.useState<Column | null>(null)
   const filterActions = React.useMemo(
@@ -35,11 +29,13 @@ export function Provider({ children }: { children: React.ReactNode }) {
     [],
   )
 
-  const [hiddenColumns, setHiddenColumns] = React.useState<HiddenColumns>(new Map())
+  const [hiddenColumns, setHiddenColumns] = React.useState<HiddenColumns>(
+    Eff.HashSet.empty(),
+  )
   const columnsActions = React.useMemo(
     () => ({
-      show: (f: Column['filter']) => setHiddenColumns(removeBool(f)),
-      hide: (f: Column['filter']) => setHiddenColumns(addBool(f)),
+      show: (f: Column['filter']) => setHiddenColumns(Eff.HashSet.remove(f)),
+      hide: (f: Column['filter']) => setHiddenColumns(Eff.HashSet.add(f)),
     }),
     [],
   )
diff --git a/catalog/app/containers/Search/Table/Table.tsx b/catalog/app/containers/Search/Table/Table.tsx
index 94b7196b..2ea3ca02 100644
--- a/catalog/app/containers/Search/Table/Table.tsx
+++ b/catalog/app/containers/Search/Table/Table.tsx
@@ -1,4 +1,5 @@
 import cx from 'classnames'
+import * as Eff from 'effect'
 import invariant from 'invariant'
 import * as React from 'react'
 import * as M from '@material-ui/core'
@@ -937,7 +938,7 @@ function ConfigureColumns({ columns, state }: ConfigureColumnsProps) {
     <>
       <ConfigureColumnsButton
         className={classes.root}
-        hasHidden={!!hiddenColumns.size}
+        hasHidden={!!Eff.HashSet.size(hiddenColumns)}
         onClick={show}
       />
 
diff --git a/catalog/app/containers/Search/Table/useColumns.ts b/catalog/app/containers/Search/Table/useColumns.ts
index 3c3cf639..62c2de1a 100644
--- a/catalog/app/containers/Search/Table/useColumns.ts
+++ b/catalog/app/containers/Search/Table/useColumns.ts
@@ -1,4 +1,5 @@
 import * as React from 'react'
+import * as Eff from 'effect'
 
 import * as JSONPointer from 'utils/JSONPointer'
 
@@ -186,7 +187,7 @@ export function useColumns(
     (): ColumnSystemMeta =>
       ColumnNameCreate({
         filtered: !!modifiedFilters && !!modifiedFilters.name,
-        visible: !hiddenColumns.has('name'),
+        visible: !Eff.HashSet.has(hiddenColumns, 'name'),
         inferred: false,
       }),
     [hiddenColumns, modifiedFilters],
@@ -196,7 +197,7 @@ export function useColumns(
     (): ColumnBucket =>
       ColumnBucketCreate({
         filtered: !!state.buckets.length,
-        visible: !hiddenColumns.has('bucket'),
+        visible: !Eff.HashSet.has(hiddenColumns, 'bucket'),
         inferred: false,
       }),
     [hiddenColumns, state.buckets.length],
@@ -208,7 +209,7 @@ export function useColumns(
       return ColumnSystemMetaCreate(
         {
           filtered: !!modifiedFilters && !!modifiedFilters[filter],
-          visible: !!predicate && !hiddenColumns.has(filter),
+          visible: !!predicate && !Eff.HashSet.has(hiddenColumns, filter),
           inferred: false,
         },
         filter,
@@ -228,7 +229,7 @@ export function useColumns(
       ColumnUserMetaCreate(
         {
           filtered: !!modifiedUserMetaFilters?.find(({ path }) => path === filter),
-          visible: !hiddenColumns.has(filter),
+          visible: !Eff.HashSet.has(hiddenColumns, filter),
           inferred,
         },
         filter,
```

fiskus replied:
Feel free to move this discussion into Slack, https://github.com/quiltdata/specs/ or in a separate PR here in quilt

nl0 commented on catalog/app/containers/NavBar/NavBar.tsx:
```diff
@@ -141,16 +143,20 @@ export function Header({ children }: HeaderProps) {
-            <M.Container className={classes.container} maxWidth="lg">
+            <LayoutContainer
+              className={classes.container}
+              fullWidth={searchUIModel?.state.view === SearchUIModel.View.Table}
```
this seems like a crude ad-hoc workaround for the problem.
here's an example implementation of a value overrides from anywhere in the component tree: https://github.com/quiltdata/quilt/pull/4434

fiskus replied:
Used `Provider` pattern to set `Layout.useFullWidth(true)` in Table view

nl0 replied:
> Used Provider pattern to set Layout.useFullWidth(true) in Table view

that's a step forward, but the implementation is error-prone (especially in the case when several components try to override this state), because it relies on imperative state changes.

nl0 commented on catalog/package.json:
```diff
@@ -54,6 +54,7 @@
     "@finos/perspective-viewer-datagrid": "^1.9.4",
     "@material-ui/core": "^4.11.4",
     "@material-ui/data-grid": "^4.0.0-alpha.27",
+    "@material-ui/icons": "^4.11.3",
```
i did some research and here's what i found:
1. outlined version of the visibility icon (and many others) is located in a separate font (Material Icons Outlined), and our (old) version of MUI does not support that out of the box. in order to use it we have to: a) load this font (add it to the querystring of the font loading rule), and b) manually set the class to `material-icons-outlined`.
2. icon fonts are considered bad practice for quite a while (~10 years lol), and generally we should use SVG if/when possible.

considering all that, i think we should use these SVG icons in all new code, and gradually migrate away from font icons

fiskus replied:
Is this thread resolved?

nl0 replied:
yep

nl0 commented on catalog/app/components/Layout/Container.tsx:
```diff
@@ -16,23 +16,15 @@ export function Provider({ children }: ProviderProps) {
-
-  return fullWidth
+    setFullWidth(true)
+    return () => setFullWidth(false)
```
if you use this in two different components, and one of them unmounts, state will be set to false -- that doesn't seem like intended behavior

fiskus replied:
My intention was to use this hook/setter only once on the page. This is a global per-page flag, so, I don't expect it to be used differently. Nevertheless, I've added check that this hook called only once.

fiskus commented on catalog/app/containers/Search/Table/useColumns.ts:
```diff
@@ -0,0 +1,176 @@
+    if (infered instanceof Error || infered === Workflow.Loading) return infered
+    const list = Object.keys(infered.workflow).length ? infered.workflow : infered.all
+    for (const filter in list) {
+      output.push({
```
Converted to `Map`

fiskus commented on catalog/app/containers/Search/List/index.tsx:
```diff
@@ -0,0 +1,224 @@
+        case 'InvalidInput':
+          const [err] = r.data.errors
+          const kind = err.name === 'QuerySyntaxError' ? 'syntax' : 'unexpected'
+          const details =
```
Partly refactored

fiskus commented on catalog/app/containers/Bucket/Bucket.tsx:
```diff
@@ -0,0 +1,158 @@
+      latestOnly: true,
+    }),
+    [bucket],
+  )
```
If `location.search` is relevant, then this initialization is ignored

However, instead of using `fromURLSearchParams` I use `PackagesSearchFilterIO.initialState` and `new UserMetaFilters()`

fiskus commented on catalog/app/containers/Search/Table/Entries.tsx:
```diff
@@ -0,0 +1,357 @@
+              data={data}
+              noDownload={undefined}
+              onData={undefined}
+              props={undefined} // these props go to the render functions
```
Fixed

fiskus commented on catalog/app/containers/Search/Layout/LoadNextPage.tsx:
```diff
@@ -0,0 +1,43 @@
+  },
+}))
+
+interface LoadNextPageProps {
+  className: string
+  loading?: boolean
```
I prefer to keep it required

fiskus commented on catalog/app/containers/Search/Layout/MoreButton.tsx:
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react'
+import * as M from '@material-ui/core'
+
+const useMoreButtonStyles = M.makeStyles({
+  title: {
+    paddingLeft: '3px',
+  },
+})
```
It is an attempt to align icons of two very different components with some hardcoded margins (MUI has -2px margin for icon, for example).

fiskus commented on catalog/app/containers/Search/Table/Table.tsx:
```diff
@@ -0,0 +1,1300 @@
+function FilterDialogSystemMeta({ column, onClose }: FilterDialogSystemMetaProps) {
+  const model = SearchUIModel.use(SearchUIModel.ResultType.QuiltPackage)
+
+  const [innerState, setInnerState] = React.useState<$TSFixMe>(null)
```
It's difficult to figure out, what type to pass, exactly.
Probably, need to pass some generic `FilterDialogSystemMeta`

fiskus commented on catalog/app/containers/Search/FilterWidget.tsx:
```diff
@@ -0,0 +1,223 @@
+}
+
+export default function FilterWidget(props: FilterWidgetProps) {
+  const Widget = WIDGETS[props.state._tag]
+  return <Widget {...(props as $TSFixMe)} />
```
It was in original code, I don't know how to fix this

fiskus commented on catalog/app/containers/Search/NoResults.tsx:
```diff
@@ -3,21 +3,23 @@ import * as React from 'react'
 
-export function ResultsSkeleton({ className, type }: ResultsSkeletonProps) {
+export function Skeleton({ className, state }: SkeletonProps) {
+  if (state.view === SearchUIModel.View.Table) return <TableSkeleton />
```
I think, these `className`s are not used, actually.

fiskus commented on catalog/app/components/Layout/Container.tsx:
```diff
@@ -0,0 +1,81 @@
+  React.useEffect(() => {
+    invariant(
+      !counter.current,
+      'Hook intended to set `fullWidth: true` only once. If you need a different behaviour, please, consider using different implementation.',
```
Thanks!

fiskus commented on catalog/app/containers/Search/FilterWidget.tsx:
```diff
@@ -0,0 +1,223 @@
+  const debouncedCallback = useDebouncedCallback(onChange, delay)
+
+  React.useEffect(() => {
+    if (!debouncedCallback.isPending()) setValue(initialValue)
```
I don't know, but I guess it is

fiskus commented on catalog/app/containers/Search/NoResults.tsx:
```diff
@@ -133,9 +141,11 @@ export function EmptyResults({ className }: EmptyResultsProps) {
             Reset the <StyledLink onClick={clearFilters}>search filters</StyledLink>
           </li>
         )}
+        {/* TODO:
```
Yes, if I have time for this

nl0 commented:
a layout suggestion:
"create package" button feels somewhat out of place near the result count --
maybe move it to the right end of the tab bar?

<img width="1395" height="589" alt="Screenshot 2025-07-14 at 17 14 28" src="https://github.com/user-attachments/assets/3445e0f4-76ff-4f49-98b3-72cc39fc2a8e" />

fiskus commented on catalog/app/containers/Search/Table/useColumns.ts:
```diff
@@ -0,0 +1,275 @@
+            // and we don't need to keep fillng `allFacets`
+            if (workflowFacets.size) return
+
+            // Use keywords when possible
```
It depends on "keywords" implementation. If values become "keywords" when there is only a small variety of them, then I would prefer a list of these values to choose from, rather than slider or text field.

fiskus commented on catalog/app/containers/Search/model.ts:
```diff
@@ -1548,4 +1617,47 @@ export function useSearchUIModelContext(type?: ResultType) {
+  }
+}
+
+export const useMakeObjectsSearchUrlState = (overrides: Partial<ObjectsSearchUrlState>) =>
```
This is not used. Similar `useMakePackagesSearchUrlState` is used

fiskus commented on catalog/app/containers/Search/model.ts:
```diff
@@ -609,14 +636,17 @@ function serializeSearchUrlState(state: SearchUrlState): URLSearchParams {
 }
 
-function useUrlState(): SearchUrlState {
+function useUrlState(optUrlState?: SearchUrlState): SearchUrlState {
```
1. Renamed to `defaults`
2. Value changed from `SearchUrlState` to `Partial<SearchUrlState>`

Properties from this default state are used when user not set them explicitly

fiskus commented on catalog/app/containers/Search/model.ts:
```diff
@@ -1263,13 +1317,14 @@ export function usePackageUserMetaFacetExtents(path: string): {
 
-function useSearchUIModel() {
-  const urlState = useUrlState()
+function useSearchUIModel(optUrlState?: SearchUrlState, optBase?: string) {
```
Changed to defaults

fiskus commented:
* Fixed showing syntax error
* Moved search suggestions to `OpenLanding/Search`
* Moved `useRequest` to `utils/useRequest`, and added `Idle` state
* Renamed ElasticSearch version to 6.8
* Fixed user meta:
  + fix showing results after filtering user meta, that weren't in the first chunk
  + use additional hook to load user meta facets

nl0 reviewed (dismissed):
looks good. some nitpicks inline, plus some ui quirks/suggestions posted in slack

fiskus commented on catalog/app/containers/Search/model.ts:
```diff
@@ -576,18 +612,37 @@ export function parseSearchParams(qs: string): SearchUrlState {
 }
 
+function areBucketsEqual(left: readonly string[], right: readonly string[]) {
+  if (!left.length && !right.length) return true
+  if (left.length !== right.length) return false
+  const leftCache: Record<(typeof left)[number], null> = left.reduce(
+    (memo, l) => ({ ...memo, [l]: true }),
+    {},
+  )
+  return right.every((r) => leftCache[r])
```
You're right, I forgot to sort: `R.equals(left.sort(), right.sort())`

> it is already in the project, so why not use it?

Because, you didn't answer the question of why use it in the first place.

I explicitly didn't commit to use this library.

Library is in the project, because we decided that the PR introduced it must be merged ASAP, and I didn't have time to discuss this or object.

nl0 replied:
> You're right, I forgot to sort: R.equals(left.sort(), right.sort())

this will sort arrays in place, mutating them, which is not quite desirable.

anyways, if it is that simple, why did you have to write a complex implementation?

> Because, you didn't answer the question of why use it in the first place.
> Library is in the project, because we decided that the PR introduced it must be merged ASAP, and I didn't have time to discuss this or object.

it provides a powerful set of tools for general-purpose typescript development,
and that allowed me to implement qurator in a safe (in terms of type checks and correctness)
and efficient (in terms of developer productivity and not reinventing wheels) way.
but its extent is far broader than qurator.

> I explicitly didn't commit to use this library.

so you're just going to ignore it and reimplement functionality it provides just because you don't like it and you did not choose it?
okay, whatever, i'm tired of arguing

nl0 commented on catalog/CHANGELOG.md:
```diff
@@ -18,6 +18,8 @@ where verb is one of
 
 ## Changes
 
+- [Added] Add table grid to show search results (including matched enties) ([#4413](https://github.com/quiltdata/quilt/pull/4413))
```
```suggestion
- [Added] Search: Table view for package search results (including matching entries) ([#4413](https://github.com/quiltdata/quilt/pull/4413))
```

nl0 commented on catalog/CHANGELOG.md:
```diff
@@ -18,6 +18,8 @@ where verb is one of
 ## Changes
 
+- [Added] Add table grid to show search results (including matched enties) ([#4413](https://github.com/quiltdata/quilt/pull/4413))
+- [Changed] Replace packages list with search results on Packages tab ([#4413](https://github.com/quiltdata/quilt/pull/4413))
```
```suggestion
- [Changed] Packages tab: Use search view to navigate packages ([#4413](https://github.com/quiltdata/quilt/pull/4413))
```
